### PR TITLE
Seed monitoring enhancement

### LINKED ISSUE #4683: kube-state-metrics for seed components
**How to categorize this issue?**

/area monitoring
/kind enhancement

**What would you like to be added**:

Scrape and keep kube-state-metrics for seed components (gardenlet, etcd-druid, extensions, etc.), i.e. everything that is not in a shoot namespace and is not already contained in the shoot monitoring stack.
Should be added to the new seed monitoring stack: https://github.com/gardener/gardener/pull/4560

**Why is this needed**:

Operators need to know what's going on with seed components, e.g. rolling updates, pod resource requests and limits, etc.

### PR DESCRIPTION
**Pull request by wyb1**
**How to categorize this PR?**

/area monitoring
/kind enhancement
/kind bug

**What this PR does / why we need it**:

- Fixes an issue when scraping the gardenlet. Https is now used instead of http. Original PR #4560. Partial fix for #2815 

- Add kube-state-metrics metrics to the seed-prometheus #4683

- Seed prometheus scrapes itself for meta information

**Which issue(s) this PR fixes**:
 Fixes #4683

**Release note**:

```bugfix operator
Fixes an issue when scraping the gardenlet. Https is now used instead of http.
```
```other operator
Add kube-state-metrics metrics to the seed-prometheus
```
```other operator
Seed prometheus scrapes itself for meta information
```

### CODE REVIEW

File: charts/seed-bootstrap/charts/monitoring/templates/config.yaml [hunk 1]
```diff
@@ -32,6 +32,9 @@ data:
         regex: __meta_kubernetes_pod_label_(.+)
 
     - job_name: garden
+      scheme: https
```
Comment (timebertt) [inline charts/seed-bootstrap/charts/monitoring/templates/config.yaml:35]: Can we be sure, that all scraped components support https?
Should we maybe have one rule for https (default) and one for http (pods labelled with `prometheus.io/scheme=http`) or something similar?

File: charts/seed-bootstrap/charts/monitoring/templates/config.yaml [hunk 2]
```diff
@@ -68,6 +68,30 @@ data:
...
+      kubernetes_sd_configs:
+      - role: service
+        namespaces:
+          names: [ garden ]
+      relabel_configs:
+      - source_labels: [ __meta_kubernetes_service_label_component ]
+        action: keep
+        regex: kube-state-metrics
+      - source_labels: [ __meta_kubernetes_service_port_name ]
+        action: keep
```
Comment (timebertt) [inline charts/seed-bootstrap/charts/monitoring/templates/config.yaml]: Can you explain the purpose of this relabel config? Is the regex for port name `metrics` missing?

File: charts/seed-bootstrap/charts/monitoring/templates/config.yaml [hunk 3]
```diff
@@ -68,6 +68,30 @@ data:
...
+      - target_label: instance
+        replacement: kube-state-metrics
+      metric_relabel_configs:
+      - source_labels: [ pod ]
+        regex: ^.+\.tf-pod.+$
+        action: drop
```
Comment (timebertt) [inline charts/seed-bootstrap/charts/monitoring/templates/config.yaml]: I guess, there shouldn't be any `tf-pod` metrics in the non-shoot namespaces, right?

File: charts/seed-bootstrap/charts/monitoring/templates/config.yaml [hunk 4]
```diff
@@ -68,6 +68,30 @@ data:
...
+      - source_labels: [ namespace ]
+        regex: ^shoot--.+$
```
Comment (timebertt) [inline charts/seed-bootstrap/charts/monitoring/templates/config.yaml]: The shoot technical ID used to start with `shoot-` (only one hyphen). We should consider this to filter out metrics from namespaces of very old shoots.

**Review by timebertt — CHANGES_REQUESTED** [review]
Thank you very much, looking forward to it!

File: charts/seed-bootstrap/charts/monitoring/templates/config.yaml (hunk 2 above)
Comment (wyb1) [reply charts/seed-bootstrap/charts/monitoring/templates/config.yaml]: This config will keep the metrics from the target as long as it has a named port. It is probably better to explicitly add a regex to match the `metrics` port though.

File: charts/seed-bootstrap/charts/monitoring/templates/config.yaml (hunk 3 above)
Comment (wyb1) [reply charts/seed-bootstrap/charts/monitoring/templates/config.yaml]: I wasn't entirely sure so I kept the relabel configs, but if we don't have any tf-pods in non-shoot namespaces I will remove this so we don't have unnecessary config.

File: charts/seed-bootstrap/charts/monitoring/templates/config.yaml (hunk 1 above)
Comment (wyb1) [reply charts/seed-bootstrap/charts/monitoring/templates/config.yaml:35]: We thought about this, but wanted to keep the config as is until we need to support this (which will probably happen very soon). If you prefer I can add this config now.

Comment (timebertt) [reply charts/seed-bootstrap/charts/monitoring/templates/config.yaml:35]: Fair enough, let's add it later once we need it 👍

**Comment by wyb1** [comment]
Thanks for the review! Addressed the comments. PTAL

File: charts/seed-bootstrap/charts/monitoring/templates/config.yaml [hunk 5]
```diff
@@ -68,6 +68,28 @@ data:
...
+        regex: kube-state-metrics
+        action: keep
+      - source_labels: [ __meta_kubernetes_service_port_name ]
+        regex: metrics
+        action: keep
+      - target_label: instance
+        replacement: kube-state-metrics
+      metric_relabel_configs:
+      - source_labels: [ namespace ]
+        regex: shoot-.+
```
Comment (timebertt) [inline charts/seed-bootstrap/charts/monitoring/templates/config.yaml:90]: Why did you remove the start/end tokens (`^...$`)? Does this match `extension-extension-shoot-dns-service-abcdef` now?

**Comment by timebertt** [comment]
@wyb1 BTW, feel free to push new commits for addressing review feedback instead of force-pushing amended commits. That makes it easier for reviewers to spot the diff between several reviews, especially for larger changes. Doesn't matter to much for this PR though, as it's quite small :)

File: charts/seed-bootstrap/charts/monitoring/templates/config.yaml (hunk 5 above)
Comment (wyb1) [reply charts/seed-bootstrap/charts/monitoring/templates/config.yaml:90]: Prometheus automatically anchors the regex. So it will not match the namespace that you mentioned. https://prometheus.io/docs/prometheus/latest/configuration/configuration/#regex

Comment (timebertt) [reply charts/seed-bootstrap/charts/monitoring/templates/config.yaml:90]: Ok, perfect! 👍

**Review by timebertt — APPROVED** [review]
Thanks again!
/lgtm

**Comment by timebertt** [comment]
@wyb1 I think it would be nice to backport the fix for scraping the gardenlet to `release-v1.32`. WDYT?

**Comment by wyb1** [comment]
@timebertt I agree. I will do the cherry picks later today

**Comment by wyb1** [comment]
Cherry picked for `release-v1.32` 
https://github.com/gardener/gardener/pull/4716
